warp-tls/Network/Wai/Handler/WarpTLS.hs

```diff
@@ -276,15 +302,19 @@ runTLS tset set app = withSocketsDo $
 
 loadCredentials :: TLSSettings -> IO TLS.Credentials
 loadCredentials TLSSettings{ tlsCredentials = Just creds } = return creds
-loadCredentials TLSSettings{..} = case (certMemory, keyMemory) of
-    (Nothing, Nothing) -> do
-        cred <- either error id <$> TLS.credentialLoadX509Chain certFile chainCertFiles keyFile
-        return $ TLS.Credentials [cred]
-    (mcert, mkey) -> do
-        cert <- maybe (S.readFile certFile) return mcert
-        key <- maybe (S.readFile keyFile) return mkey
-        cred <- either error return $ TLS.credentialLoadX509ChainFromMemory cert chainCertsMemory key
-        return $ TLS.Credentials [cred]
+loadCredentials TLSSettings{..} = case certSettings of 
+  CertFromFile cert chainFiles key -> do
+    cred <- either error id <$> TLS.credentialLoadX509Chain cert chainFiles key
+    return $ TLS.Credentials [cred]
+  CertFromRef certRef chainCertsRef keyRef -> do 
+    cert <- I.readIORef certRef
+    chainCerts <- mapM I.readIORef chainCertsRef
+    key <- I.readIORef keyRef
+    cred <- either error return $ TLS.credentialLoadX509ChainFromMemory cert chainCerts key
+    return $ TLS.Credentials [cred]
```

Comment:
Which brings us to perhaps the most important issue. What is the purpose of the `IORef` alternative, if ultimately `loadCredentials` converts them (if I'm not mistaken just once) immediately to a set of pure objects used in:
```haskell
runTLSSocket :: TLSSettings -> Settings -> Socket -> Application -> IO ()
runTLSSocket tlsset set sock app = do
    credentials <- loadCredentials tlsset
    mgr <- getSessionManager tlsset
    runTLSSocket' tlsset set credentials mgr sock app
```
This still means that the application loop needs to be restarted when new certificates are available, and so I fail to see the difference between this `CertFromMemory`.  What does the indirection via `IORef`s achieve?

It seems like there's much missing code here to make a practically useful deployment model.  Perhaps the idea is to also provide a restart loop (I may start work on that shortly) and have the restart loop be configured once with these (or ideally just one) `IORefs`, allowing another thread to just worry about updating the data as needed?

If so I think that code should probably have appeared concurrently with or even before this PR, which would have made it easier to see whether this API is the right one for the job.

[ My tentative loop restart implementation idea will only work on POSIX systems that support dup(2) on socket file descriptors. Which precludes Windows.  I hope that having reasonably clean certificate reload without dropping any connections even briefly, but just on Unix-like systems is an acceptable design choice.  More ambitious designs would likely require invasive changes in the TLS library, which are perhaps a good idea, but I'm not planning to go there at this time. ]